From 4844bf3571184f7cf45885bb818126d81e9918a3 Mon Sep 17 00:00:00 2001
From: Karthikeyan K <karthikk@habana.ai>
Date: Wed, 21 Apr 2021 08:52:22 +0300
Subject: [PATCH] [SW-40084] greco: efuse: read address range fix

Currently the read address register(PSOC_EFUSE_READ_ADDRESS)
for efuse contains a value of 0x3FF8 on reset, which prevents
firmware from reading efuse contents from 0x00 - 0x77.

This value is set by the boot manager during the boot flow.

In order to read all the contents of efuse the
value needs to be updated to 0x3F80 for legacy boot flow

In case of secure boot flow, it is done in arcbootrom

This patch contains the fix for the same.

Change-Id: I159fa0fa58ffa4bf6bd02b94597400562980d7d9
Signed-off-by: Karthikeyan K <karthikk@habana.ai>
---

diff --git a/common/efuse.c b/common/efuse.c
index b2088ff..f3552b7 100644
--- a/common/efuse.c
+++ b/common/efuse.c
@@ -28,6 +28,8 @@
 
 static int efuse_init;
 
+__weak void preboot_efuse_read_range_set(void) {}
+
 int preboot_fuse_init(void)
 {
 	int i;
@@ -38,6 +40,8 @@
 
 	read_reg = get_efuse_read_reg();
 
+	preboot_efuse_read_range_set();
+
 	regmap_writel(1, read_reg);
 
 	for (i = 0; i < FUSE_RETRIES; i++)
diff --git a/greco/common.c b/greco/common.c
index a53324d..7011301 100644
--- a/greco/common.c
+++ b/greco/common.c
@@ -44,12 +44,18 @@
 #include "common.h"
 
 #define FW_WORD_NUM		0x2000	/* FW_SIZE / 2 */
+#define FUSE_READ_RANGE_ALL	0x3F80
 
 u32 get_efuse_read_reg(void)
 {
 	return mmPSOC_EFUSE_READ;
 }
 
+void preboot_efuse_read_range_set(void)
+{
+	regmap_writel(FUSE_READ_RANGE_ALL, mmPSOC_EFUSE_READ_ADDRESS);
+}
+
 bool hl_is_debug_mode(enum hl_debug_hw_func func)
 {
 	u32 boot_strap = regmap_readl(mmPSOC_GLOBAL_CONF_BOOT_STRAP_PINS_L);
